examples/controlnet/train_controlnet.py

```diff
@@ -0,0 +1,1179 @@
+from optimum.habana.utils import set_seed
+
+
+try:
+    from optimum.habana.utils import check_optimum_habana_min_version
+except ImportError:
+
+    def check_optimum_habana_min_version(*a, **b):
+        return ()
+
+
+# Will error if the minimal version of Optimum Habana is not installed. Remove at your own risks.
+check_optimum_habana_min_version("1.8.1")
+if is_wandb_available():
+    import wandb
+
+# Will error if the minimal version of diffusers is not installed. Remove at your own risks.
+check_min_version("0.23.0")
+
+logger = get_logger(__name__)
+
+
+def image_grid(imgs, rows, cols):
+    assert len(imgs) == rows * cols
+
+    w, h = imgs[0].size
+    grid = Image.new("RGB", size=(cols * w, rows * h))
+
+    for i, img in enumerate(imgs):
+        grid.paste(img, box=(i % cols * w, i // cols * h))
+    return grid
+
+
+def log_validation(
+    vae, text_encoder, tokenizer, unet, controlnet, args, accelerator, noise_scheduler, weight_dtype, step
+):
+    logger.info("Running validation... ")
+
+    controlnet = accelerator.unwrap_model(controlnet)
+
+    pipeline = GaudiStableDiffusionControlNetPipeline.from_pretrained(
+        args.pretrained_model_name_or_path,
+        vae=vae,
+        text_encoder=text_encoder,
+        tokenizer=tokenizer,
+        unet=unet,
+        controlnet=controlnet,
+        safety_checker=None,
+        revision=args.revision,
+        variant=args.variant,
+        scheduler=noise_scheduler,
+        use_habana=True,
+        use_hpu_graphs=True,
+        gaudi_config=args.gaudi_config_name,
+    )
+    pipeline.scheduler = UniPCMultistepScheduler.from_config(pipeline.scheduler.config)
+    pipeline = pipeline.to(accelerator.device)
+    pipeline.set_progress_bar_config(disable=True)
+
+    if args.seed is None:
+        generator = None
+    else:
+        generator = torch.Generator(device=accelerator.device).manual_seed(args.seed)
+
+    if len(args.validation_image) == len(args.validation_prompt):
+        validation_images = args.validation_image
+        validation_prompts = args.validation_prompt
+    elif len(args.validation_image) == 1:
+        validation_images = args.validation_image * len(args.validation_prompt)
+        validation_prompts = args.validation_prompt
+    elif len(args.validation_prompt) == 1:
+        validation_images = args.validation_image
+        validation_prompts = args.validation_prompt * len(args.validation_image)
+    else:
+        raise ValueError(
+            "number of `args.validation_image` and `args.validation_prompt` should be checked in `parse_args`"
+        )
+
+    image_logs = []
+
+    for validation_prompt, validation_image in zip(validation_prompts, validation_images):
+        validation_image = Image.open(validation_image).convert("RGB")
+
+        images = []
+
+        for _ in range(args.num_validation_images):
+            image = pipeline(validation_prompt, validation_image, num_inference_steps=20, generator=generator).images[
```

Comment:
original code seems running this under autocast, reason why this is removed?

Reply:
The autocast was with cuda and removed. Other changes were based on changes made to textual_inversion.py which has no autocast with hpu comparing: https://github.com/huggingface/optimum-habana/blob/main/examples/stable-diffusion/textual_inversion.py#L140 vs https://github.com/huggingface/diffusers/blob/main/examples/textual_inversion/textual_inversion.py#L139

Reply:
https://github.com/huggingface/diffusers/blob/main/examples/controlnet/train_controlnet.py#L128 
Here I still see autocast.   ?? 

Can you try with autocast(bf16) and see if you are seeing any accuracy issue compare to fp32?

Reply:
Yes the autocast is with cuda. I guess I am trying to say for textual_inversion it was not carried over. Yes I can do the suggested experiment and compare.

Reply:
Will use torch.autocast here, experiment reveals positive impact on perf and no impact to accuracy.